### DISCUSSION TITLE
[Bug]: 28.10.3 regression and getting ERR_UNSUPPORTED_ESM_URL_SCHEME error in serverless functions.

### DISCUSSION THREAD

**Discussion by user72356** [vercel/vercel#9283] [Help] [answered]
UPDATE: I'm moving to Netlify because it's clear anything non-Next.js at Vercel is a distant second-class citizen. I've had nothing but issues, which have been reported and never resolved. First impressions of Netlify is that it's everything good about Vercel, without the bugs.

Vercel CLI version 28.10.3 introduced a regression and my pure Vercel serverless functions (not Next.js), which are written in Typescript, no longer work, in dev or prod. I'm getting an ERR_UNSUPPORTED_ESM_URL_SCHEME error. It works fine for serverless function written in Javascript. This is with CRA/Vite projects, node versions 16/18.

The scary thing is that this version of vercel CLI has been out for weeks and no one has noticed/fixed this obvious bug, casting doubt about the level of attention non-Next.js serverless gets.

Simple steps to reproduce. It doesn't get simpler:

- create-react-app with Typescript template (or create-vite if you prefer Vite)
- Add an api folder and a serverless function written in Typescript that returns "Hello World"
- Execute vercel dev
- Go to your hello world function on localhost and enjoy your ERR_UNSUPPORTED_ESM_URL_SCHEME error.
- Push to prod and enjoy the error in prod.

```
$ vercel dev
Vercel CLI 28.10.3
> Running Dev Command “vite --port $PORT”

  VITE v4.0.4  ready in 403 ms

  ➜  Local:   http://localhost:3000/
  ➜  Network: use --host to expose
  ➜  press h to show help
> Ready! Available at http://localhost:3000
(node:23556) ExperimentalWarning: Custom ESM Loaders is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
node:internal/errors:491
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only URLs with a scheme in: file, data are supported by the default ESM loader. On Windows, absolute paths must be valid file:// URLs. Received protocol 'c:'
    at new NodeError (node:internal/errors:400:5)
    at throwIfUnsupportedURLScheme (node:internal/modules/esm/resolve:1055:11)
    at defaultResolve (node:internal/modules/esm/resolve:1135:3)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:842:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ESMLoader.import (node:internal/modules/esm/loader:525:22)
    at initializeLoader (node:internal/process/esm_loader:75:58)
    at loadESM (node:internal/process/esm_loader:90:11)
    at runMainESM (node:internal/modules/run_main:55:21) {
  code: 'ERR_UNSUPPORTED_ESM_URL_SCHEME'
}

Node.js v18.13.0
Error: Function `api/test/index.ts` failed with exit code 1
```

**Comment by amyegan** [comment, marked as answer]
A change was recently released to support ESM with TS (https://github.com/vercel/vercel/pull/7954). In another discussion, a few solutions were shared to handle failing configurations after the change was released: https://github.com/vercel/community/discussions/1225#discussioncomment-4642381. The error message you're seeing is different, but the timing seems to be right for this code change. 

I also found some suggestions on The Linux Foundation forum ([Lab 7.1 ERR_UNSUPPORTED_ESM_URL_SCHEME](https://forum.linuxfoundation.org/discussion/861047/lab-7-1-err-unsupported-esm-url-scheme)) and Stack Overflow ([Only file and data URLs are supported by the default ESM loader - Vue 3](https://stackoverflow.com/questions/69665780/error-err-unsupported-esm-url-scheme-only-file-and-data-urls-are-supported-by))

If none of these solutions work in your situation, please share the package.json, tsconfig.json, runtime error message (or deployment URL), and the contents of the file that failed. If you have a public repo or [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), a link to that can be used instead of individual file contents. 🙏
